File: vol2/ch21/LeetCode337-1.py
class Solution:
  def rob(self, root: Optional[TreeNode]) -> int:
    self.hmap={}
    if root==None:return 0
    return self.dfs(root)
  def dfs(self,root):
    if not root:return 0
    if root in self.hmap: #该子问题已经求出，直接返回解
      return self.hmap[root]
    money1=root.val
    if root.left!=None:
      money1+=self.dfs(root.left.left)+self.dfs(root.left.right)
    if root.right!=None:
      money1+=self.dfs(root.right.left)+self.dfs(root.right.right)
    money2=self.dfs(root.left)+self.dfs(root.right)
    self.hmap[root]=max(money1,money2)
    return self.hmap[root]

File: vol2/ch21/LeetCode337-1.cpp
class Solution {
  unordered_map<TreeNode*,int> hmap;
public:
  int rob(TreeNode* root) {
    if (root==NULL) return 0;
          return dfs(root);
      }
      int dfs(TreeNode* root) {
        if (root==NULL) return 0;
          if(hmap.find(root)!=hmap.end())     //该子问题已经求出，直接返回解
              return hmap[root];
          int money1=root->val;
          if (root->left!=NULL)
              money1+=dfs(root->left->left)+dfs(root->left->right);
          if (root->right!=NULL)
              money1+=dfs(root->right->left)+dfs(root->right->right);
          int money2=dfs(root->left)+dfs(root->right);
          hmap[root]=max(money1,money2);
          return hmap[root];
      }
};

